x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/IndexServiceAccountsTokenStore.java

```diff
@@ -0,0 +1,196 @@
+                client.prepareIndex(SECURITY_MAIN_ALIAS)
+                    .setId(docIdForToken(token))
+                    .setSource(builder)
+                    .setOpType(OpType.CREATE)
+                    .setRefreshPolicy(request.getRefreshPolicy())
+                    .request();
+            final BulkRequest bulkRequest = toSingleItemBulkRequest(indexRequest);
+
+            securityIndex.prepareIndexIfNeededThenExecute(listener::onFailure, () -> {
+                executeAsyncWithOrigin(client, SECURITY_ORIGIN, BulkAction.INSTANCE, bulkRequest,
+                    TransportSingleItemBulkWriteAction.<IndexResponse>wrapBulkResponse(ActionListener.wrap(response -> {
+                        if (DocWriteResponse.Result.CREATED == response.getResult()) {
+                            listener.onResponse(CreateServiceAccountTokenResponse.created(
+                                token.getTokenName(), token.asBearerString()));
+                        } else {
+                            listener.onResponse(CreateServiceAccountTokenResponse.notCreated());
+                        }
+                    }, listener::onFailure)));
+            });
+        } catch (IOException e) {
+            listener.onFailure(e);
+        }
+    }
+
+    @Override
+    public void findTokensFor(ServiceAccountId accountId, ActionListener<Collection<TokenInfo>> listener) {
+        // TODO: wildcard support?
+        final BoolQueryBuilder query = QueryBuilders.boolQuery()
+            .filter(QueryBuilders.termQuery("doc_type", SERVICE_ACCOUNT_TOKEN_DOC_TYPE))
+            .must(QueryBuilders.prefixQuery("name", accountId.asPrincipal()));
+        final SearchRequest request = client.prepareSearch(SECURITY_MAIN_ALIAS)
+            .setScroll(DEFAULT_KEEPALIVE_SETTING.get(getSettings()))
+            .setQuery(query)
+            .setSize(1000)
+            .setFetchSource(false)
+            .request();
+        request.indicesOptions().ignoreUnavailable();
+
+        logger.trace("Searching tokens for service account [{}]", accountId);
+        ScrollHelper.fetchAllByEntity(client, request,
+            new ContextPreservingActionListener<>(client.threadPool().getThreadContext().newRestorableContext(false), listener),
+            hit -> extractTokenInfo(hit.getId(), accountId));
+    }
+
+    private String docIdForToken(ServiceAccountToken token) {
+        return SERVICE_ACCOUNT_TOKEN_DOC_TYPE + "-" + token.getQualifiedName();
+    }
+
+    private XContentBuilder newDocument(Authentication authentication, ServiceAccountToken serviceAccountToken) throws IOException {
+        final Version version = clusterService.state().nodes().getMinNodeVersion();
+
+        XContentBuilder builder = XContentFactory.jsonBuilder();
+        builder.startObject()
+            .field("doc_type", SERVICE_ACCOUNT_TOKEN_DOC_TYPE)
+            .field("version", version.id)
+            .field("name", serviceAccountToken.getQualifiedName())
```

Comment:
Did you consider splitting this and having
```suggestion
            .field("username", serviceAccountToken.getAccountId().getPrincipal())
            .field("name", serviceAccountToken.getTokenName())
```
(please excuse if those method names aren't right, you know what I mean)

Reply:
Sure I think it makes more sense.